File: 백준/Gold/9251. LCS/LCS.py
import sys
input = sys.stdin.readline

row = list(input().strip())
col = list(input().strip())

dp = [[0 for _ in range(len(col)+1)] for _ in range(len(row)+1)]

for i in range(1,len(row)+1):
    for j in range(1,len(col)+1):
        if(row[i-1] == col[j-1]):
            dp[i][j] = dp[i-1][j-1] +1
        else:
            dp[i][j] = max(dp[i-1][j],dp[i][j-1])

print(max(map(max,dp)))

File: 백준/Gold/9251. LCS/LCS.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const row = input[0].split("")
const col = input[1].split("")
const dp = Array.from({ length: row.length + 1 }, () =>
  Array.from({ length: col.length + 1 }, () => 0),
)

for(let i=1;i<=row.length;i+=1){
  for(let j=1;j<=col.length;j+=1){
    if(row[i-1] === col[j-1]){
      dp[i][j] = dp[i-1][j-1]+1
    }else{
      dp[i][j] = Math.max(dp[i-1][j],dp[i][j-1])
    }
  }
}

const answer = []
dp.forEach((row)=> answer.push(...row))
console.log(Math.max(...answer))